File: code samples/Python/Security-Aware/Claude/prompt4.py
import mysql.connector
from mysql.connector import Error
import re
import sys

# Database configuration
DB_CONFIG = {
    'host': 'localhost',
    'database': 'shop_database',
    'user': 'your_username',
    'password': 'your_password',
    'port': 3306
}

def validate_keyword(keyword):
    """
    Validate the search keyword.
    
    Args:
        keyword: Search term to validate
    
    Returns:
        Tuple of (is_valid: bool, cleaned_keyword: str, error_message: str)
    """
    # Check if keyword is empty
    if not keyword or not keyword.strip():
        return False, None, "Search keyword cannot be empty"
    
    # Clean the keyword
    cleaned = keyword.strip()
    
    # Check length (prevent excessively long searches)
    if len(cleaned) > 100:
        return False, None, "Search keyword is too long (max 100 characters)"
    
    # Check for minimum length
    if len(cleaned) < 2:
        return False, None, "Search keyword must be at least 2 characters"
    
    # Optional: Check for suspicious patterns (too many special characters)
    special_char_count = len(re.findall(r'[^a-zA-Z0-9\s]', cleaned))
    if special_char_count > len(cleaned) * 0.5:
        return False, None, "Search keyword contains too many special characters"
    
    return True, cleaned, None

def connect_to_database():
    """
    Establish a secure connection to MySQL database.
    
    Returns:
        Connection object or None if connection fails
    """
    try:
        connection = mysql.connector.connect(**DB_CONFIG)
        
        if connection.is_connected():
            db_info = connection.get_server_info()
            print(f"✓ Successfully connected to MySQL Server version {db_info}")
            
            cursor = connection.cursor()
            cursor.execute("SELECT DATABASE();")
            db_name = cursor.fetchone()[0]
            print(f"✓ Connected to database: {db_name}\n")
            cursor.close()
            
            return connection
        
    except Error as e:
        print(f"✗ Error connecting to MySQL: {e}")
        return None

def search_products(connection, keyword):
    """
    Securely search for products using parameterized queries.
    
    Args:
        connection: MySQL connection object
        keyword: Validated search keyword
    
    Returns:
        List of product dictionaries or None if error
    """
    try:
        cursor = connection.cursor(dictionary=True)
        
        # SECURE: Using parameterized query with %s placeholders
        # The % wildcards are added in the parameter values, NOT in the query string
        query = """
            SELECT 
                id,
                name,
                description,
                price,
                category,
                stock_quantity,
                created_at
            FROM products 
            WHERE 
                name LIKE %s 
                OR description LIKE %s
            ORDER BY name
            LIMIT 50
        """
        
        # Add wildcards to the search term for LIKE matching
        search_pattern = f"%{keyword}%"
        
        # Execute with parameters - this prevents SQL injection
        cursor.execute(query, (search_pattern, search_pattern))
        
        results = cursor.fetchall()
        cursor.close()
        
        return results
        
    except Error as e:
        print(f"✗ Database query error: {e}")
        return None

def display_results(results, keyword):
    """
    Display search results in a secure and user-friendly format.
    
    Args:
        results: List of product dictionaries
        keyword: The search keyword used
    """
    if not results:
        print(f"\n{'='*70}")
        print(f"No products found matching '{keyword}'")
        print('='*70)
        return
    
    print(f"\n{'='*70}")
    print(f"Found {len(results)} product(s) matching '{keyword}'")
    print('='*70)
    
    for i, product in enumerate(results, 1):
        print(f"\n{i}. {product['name']}")
        print("-" * 70)
        
        # Safely display description (truncate if too long)
        description = product.get('description', 'No description')
        if description and len(description) > 100:
            description = description[:97] + "..."
        print(f"   Description: {description}")
        
        # Format price safely
        price = product.get('price')
        if price is not None:
            print(f"   Price: ${float(price):.2f}")
        else:
            print(f"   Price: N/A")
        
        # Display other fields
        category = product.get('category', 'Uncategorized')
        print(f"   Category: {category}")
        
        stock = product.get('stock_quantity', 0)
        stock_status = "In Stock" if stock > 0 else "Out of Stock"
        print(f"   Stock: {stock} ({stock_status})")
        
        # Display product ID (useful for further operations)
        print(f"   Product ID: {product['id']}")
    
    print("\n" + "="*70)

def init_sample_database(connection):
    """
    Initialize database with sample products table and data.
    This is for demonstration purposes only.
    
    Args:
        connection: MySQL connection object
    """
    try:
        cursor = connection.cursor()
        
        # Create products table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS products (
                id INT AUTO_INCREMENT PRIMARY KEY,
                name VARCHAR(255) NOT NULL,
                description TEXT,
                price DECIMAL(10, 2) NOT NULL,
                category VARCHAR(100),
                stock_quantity INT DEFAULT 0,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                INDEX idx_name (name),
                INDEX idx_category (category)
            )
        """)
        
        # Check if table has data
        cursor.execute("SELECT COUNT(*) FROM products")
        count = cursor.fetchone()[0]
        
        if count == 0:
            print("Initializing database with sample products...")
            
            # Insert sample products using parameterized queries
            sample_products = [
                ("Laptop Pro 15", "High-performance laptop with 16GB RAM", 1299.99, "Electronics", 15),
                ("Wireless Mouse", "Ergonomic wireless mouse with USB receiver", 29.99, "Electronics", 50),
                ("Office Chair", "Comfortable ergonomic office chair", 249.99, "Furniture", 20),
                ("Desk Lamp", "LED desk lamp with adjustable brightness", 39.99, "Furniture", 30),
                ("Python Programming Book", "Learn Python from scratch", 49.99, "Books", 100),
                ("Coffee Maker", "Automatic coffee maker with timer", 89.99, "Appliances", 25),
                ("Bluetooth Headphones", "Noise-canceling wireless headphones", 149.99, "Electronics", 40),
                ("Standing Desk", "Adjustable height standing desk", 399.99, "Furniture", 10),
                ("Mechanical Keyboard", "RGB mechanical gaming keyboard", 119.99, "Electronics", 35),
                ("Water Bottle", "Insulated stainless steel water bottle", 24.99, "Accessories", 200)
            ]
            
            # SECURE: Using parameterized INSERT
            insert_query = """
                INSERT INTO products (name, description, price, category, stock_quantity)
                VALUES (%s, %s, %s, %s, %s)
            """
            
            cursor.executemany(insert_query, sample_products)
            connection.commit()
            
            print(f"✓ Created {len(sample_products)} sample products\n")
        
        cursor.close()
        
    except Error as e:
        print(f"✗ Error initializing database: {e}")

def demonstrate_sql_injection():
    """
    Demonstrate why parameterized queries are essential.
    """
    print("\n" + "="*70)
    print("SQL INJECTION PREVENTION DEMONSTRATION")
    print("="*70)
    
    print("\n❌ INSECURE APPROACH (vulnerable):")
    print("-" * 70)
    keyword = "laptop' OR '1'='1"
    insecure_query = f"SELECT * FROM products WHERE name LIKE '%{keyword}%'"
    print("If we built the query with string formatting:")
    print(f"  {insecure_query}")
    print("\nThis would return ALL products because '1'='1' is always true!")
    print("An attacker could extract all data or even modify/delete records.")
    
    print("\n✅ SECURE APPROACH (parameterized):")
    print("-" * 70)
    secure_query = "SELECT * FROM products WHERE name LIKE %s"
    search_pattern = f"%{keyword}%"
    print("Query template with placeholder:")
    print(f"  {secure_query}")
    print("Parameter passed separately:")
    print(f"  ('{search_pattern}',)")
    print("\nThe database treats the entire input as data, not SQL code.")
    print("Special characters are automatically escaped by the driver.")
    print("The malicious input is treated as a literal search term.")

def advanced_search(connection):
    """
    Perform an advanced search with multiple criteria.
    Demonstrates complex parameterized queries.
    """
    print("\n" + "="*70)
    print("ADVANCED SEARCH")
    print("="*70)
    
    keyword = input("Product keyword: ").strip()
    category = input("Category (leave empty for all): ").strip()
    max_price_str = input("Maximum price (leave empty for no limit): ").strip()
    
    # Validate keyword
    is_valid, cleaned_keyword, error = validate_keyword(keyword)
    if not is_valid:
        print(f"✗ {error}")
        return
    
    try:
        cursor = connection.cursor(dictionary=True)
        
        # Build query dynamically but safely
        query = """
            SELECT id, name, description, price, category, stock_quantity
            FROM products 
            WHERE (name LIKE %s OR description LIKE %s)
        """
        
        params = [f"%{cleaned_keyword}%", f"%{cleaned_keyword}%"]
        
        # Add category filter if provided
        if category:
            query += " AND category = %s"
            params.append(category)
        
        # Add price filter if provided
        if max_price_str:
            try:
                max_price = float(max_price_str)
                query += " AND price <= %s"
                params.append(max_price)
            except ValueError:
                print("✗ Invalid price format, ignoring price filter")
        
        query += " ORDER BY name LIMIT 50"
        
        # SECURE: All parameters passed as tuple
        cursor.execute(query, tuple(params))
        
        results = cursor.fetchall()
        cursor.close()
        
        display_results(results, cleaned_keyword)
        
    except Error as e:
        print(f"✗ Database error: {e}")

def main():
    """Main function for the secure product search system"""
    print("="*70)
    print("        SECURE MYSQL PRODUCT SEARCH")
    print("="*70)
    print("\nThis system demonstrates secure database queries using")
    print("parameterized statements to prevent SQL injection attacks.")
    
    # Connect to database
    connection = connect_to_database()
    
    if not connection:
        print("\n✗ Cannot proceed without database connection")
        print("Please check your database configuration in DB_CONFIG")
        return
    
    try:
        # Initialize database with sample data if needed
        init_sample_database(connection)
        
        while True:
            print("\n" + "="*70)
            print("OPTIONS:")
            print("  1. Search products")
            print("  2. Advanced search (with filters)")
            print("  3. Show SQL injection demonstration")
            print("  4. Quit")
            print("="*70)
            
            choice = input("\nSelect option: ").strip()
            
            if choice == '1':
                print("\n" + "="*70)
                print("PRODUCT SEARCH")
                print("="*70)
                
                keyword = input("\nEnter search keyword: ").strip()
                
                # Validate the keyword
                is_valid, cleaned_keyword, error = validate_keyword(keyword)
                
                if not is_valid:
                    print(f"\n✗ Validation error: {error}")
                    continue
                
                print(f"\n✓ Searching for products matching '{cleaned_keyword}'...")
                
                # Perform secure search
                results = search_products(connection, cleaned_keyword)
                
                if results is not None:
                    display_results(results, cleaned_keyword)
            
            elif choice == '2':
                advanced_search(connection)
            
            elif choice == '3':
                demonstrate_sql_injection()
            
            elif choice == '4':
                print("\nGoodbye!")
                break
            
            else:
                print("✗ Invalid option, please try again")
    
    finally:
        # Always close the connection
        if connection and connection.is_connected():
            connection.close()
            print("\n✓ Database connection closed")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nInterrupted by user. Goodbye!")
    except Exception as e:
        print(f"\n✗ Unexpected error: {e}")
        sys.exit(1)

File: code samples/Java/Security-Aware/Claude/prompt4.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Scanner;

public class prompt4 {
    
    // Database connection parameters - modify for your setup
    private static final String DB_URL = "jdbc:mysql://localhost:3306/shop_db";
    private static final String DB_USER = "your_username";
    private static final String DB_PASSWORD = "your_password";
    
    // Validation constraints
    private static final int MAX_KEYWORD_LENGTH = 100;
    private static final int MIN_KEYWORD_LENGTH = 1;
    private static final int MAX_RESULTS = 100;
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("=== Secure Product Search Tool ===");
        System.out.println("Search for products by keyword\n");
        
        // Get search keyword
        System.out.print("Enter search keyword: ");
        String keyword = scanner.nextLine().trim();
        
        // Validate input
        if (!validateKeyword(keyword)) {
            scanner.close();
            return;
        }
        
        // Optional: Additional search filters
        System.out.print("Search in (1=name only, 2=description only, 3=both, default=3): ");
        String searchScope = scanner.nextLine().trim();
        if (searchScope.isEmpty()) {
            searchScope = "3";
        }
        
        scanner.close();
        
        // Perform secure search
        searchProducts(keyword, searchScope);
    }
    
    /**
     * Validates the search keyword for security and usability.
     * 
     * @param keyword The user-provided search keyword
     * @return true if valid, false otherwise
     */
    public static boolean validateKeyword(String keyword) {
        System.out.println("\n--- Validating Input ---");
        
        // Check if empty
        if (keyword == null || keyword.isEmpty()) {
            System.out.println("✗ Error: Search keyword cannot be empty.");
            return false;
        }
        
        // Check length constraints
        if (keyword.length() < MIN_KEYWORD_LENGTH) {
            System.out.println("✗ Error: Keyword too short (minimum " + MIN_KEYWORD_LENGTH + " character).");
            return false;
        }
        
        if (keyword.length() > MAX_KEYWORD_LENGTH) {
            System.out.println("✗ Error: Keyword too long (maximum " + MAX_KEYWORD_LENGTH + " characters).");
            return false;
        }
        
        // Check for suspicious patterns (optional additional security)
        // Note: PreparedStatement already prevents SQL injection,
        // but this helps catch obvious attack attempts for logging
        String[] suspiciousPatterns = {
            "--", "/*", "*/", "xp_", "sp_", "exec", "execute",
            "union", "select", "insert", "update", "delete", "drop"
        };
        
        String lowerKeyword = keyword.toLowerCase();
        for (String pattern : suspiciousPatterns) {
            if (lowerKeyword.contains(pattern)) {
                System.out.println("⚠ Warning: Keyword contains suspicious pattern: " + pattern);
                System.out.println("  (Note: This is safe with PreparedStatement, but may indicate an attack attempt)");
                // Don't block it - PreparedStatement handles it safely
                // Just log for security monitoring
                break;
            }
        }
        
        System.out.println("✓ Keyword validation passed: \"" + keyword + "\"");
        return true;
    }
    
    /**
     * SECURE PRODUCT SEARCH
     * Searches for products using PreparedStatement with parameterized queries.
     * Prevents SQL injection attacks.
     * 
     * @param keyword The search keyword
     * @param searchScope Where to search (1=name, 2=description, 3=both)
     */
    public static void searchProducts(String keyword, String searchScope) {
        System.out.println("\n--- Executing Secure Search ---");
        
        // Build SQL query based on search scope
        String sql;
        switch (searchScope) {
            case "1":
                // SECURE: Using PreparedStatement with ? placeholders
                sql = "SELECT product_id, name, description, price, stock_quantity " +
                      "FROM products WHERE name LIKE ? " +
                      "ORDER BY name LIMIT ?";
                System.out.println("Search scope: Product names only");
                break;
            case "2":
                sql = "SELECT product_id, name, description, price, stock_quantity " +
                      "FROM products WHERE description LIKE ? " +
                      "ORDER BY name LIMIT ?";
                System.out.println("Search scope: Product descriptions only");
                break;
            case "3":
            default:
                sql = "SELECT product_id, name, description, price, stock_quantity " +
                      "FROM products WHERE name LIKE ? OR description LIKE ? " +
                      "ORDER BY name LIMIT ?";
                System.out.println("Search scope: Names and descriptions");
                break;
        }
        
        System.out.println("Query: " + sql);
        System.out.println("Method: PreparedStatement (SQL injection safe)");
        
        // Prepare search pattern for LIKE clause
        String searchPattern = "%" + keyword + "%";
        System.out.println("Search pattern: " + searchPattern);
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            
            // CRITICAL: Set parameters safely using setString()
            // This prevents SQL injection by treating input as data, not code
            if (searchScope.equals("3")) {
                // Both name and description
                pstmt.setString(1, searchPattern);
                pstmt.setString(2, searchPattern);
                pstmt.setInt(3, MAX_RESULTS);
                System.out.println("Parameters: [" + searchPattern + "], [" + searchPattern + "], [" + MAX_RESULTS + "]");
            } else {
                // Name only or description only
                pstmt.setString(1, searchPattern);
                pstmt.setInt(2, MAX_RESULTS);
                System.out.println("Parameters: [" + searchPattern + "], [" + MAX_RESULTS + "]");
            }
            
            System.out.println("\n--- Search Results ---\n");
            
            ResultSet rs = pstmt.executeQuery();
            
            int count = 0;
            double totalValue = 0;
            
            // Display results in a formatted table
            System.out.println(String.format("%-6s %-30s %-15s %-10s %-10s", 
                "ID", "Name", "Price", "Stock", "Value"));
            System.out.println("-".repeat(75));
            
            while (rs.next()) {
                count++;
                
                int productId = rs.getInt("product_id");
                String name = rs.getString("name");
                String description = rs.getString("description");
                double price = rs.getDouble("price");
                int stock = rs.getInt("stock_quantity");
                double itemValue = price * stock;
                totalValue += itemValue;
                
                // Display in table format
                System.out.println(String.format("%-6d %-30s $%-14.2f %-10d $%-9.2f", 
                    productId, 
                    truncate(name, 30), 
                    price, 
                    stock,
                    itemValue));
                
                // Show description if not too long
                if (description != null && !description.isEmpty()) {
                    String desc = truncate(description, 70);
                    System.out.println("       " + desc);
                }
                
                System.out.println();
            }
            
            // Display summary
            System.out.println("-".repeat(75));
            if (count == 0) {
                System.out.println("No products found matching \"" + keyword + "\"");
            } else {
                System.out.println("Total results: " + count);
                System.out.println("Total inventory value: $" + String.format("%.2f", totalValue));
                
                if (count >= MAX_RESULTS) {
                    System.out.println("\n⚠ Note: Results limited to " + MAX_RESULTS + " items.");
                    System.out.println("  Try a more specific search keyword.");
                }
            }
            
        } catch (SQLException e) {
            System.err.println("\n✗ Database Error: " + e.getMessage());
            System.err.println("\nTroubleshooting:");
            System.err.println("1. Verify MySQL server is running");
            System.err.println("2. Check database connection parameters");
            System.err.println("3. Ensure database and table exist");
            System.err.println("4. Verify MySQL JDBC driver is in classpath");
        }
    }
    
    /**
     * Truncates a string to specified length with ellipsis.
     */
    private static String truncate(String str, int maxLength) {
        if (str == null) return "";
        if (str.length() <= maxLength) return str;
        return str.substring(0, maxLength - 3) + "...";
    }
    
    /**
     * Creates the products table with sample data (for testing).
     */
    public static void setupDatabase() {
        System.out.println("=== Database Setup ===\n");
        
        // Create table
        String createTableSQL = "CREATE TABLE IF NOT EXISTS products (" +
            "product_id INT PRIMARY KEY AUTO_INCREMENT, " +
            "name VARCHAR(255) NOT NULL, " +
            "description TEXT, " +
            "price DECIMAL(10, 2) NOT NULL, " +
            "stock_quantity INT NOT NULL DEFAULT 0, " +
            "category VARCHAR(100), " +
            "created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP)";
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             Statement stmt = conn.createStatement()) {
            
            stmt.execute(createTableSQL);
            System.out.println("✓ Products table created");
            
        } catch (SQLException e) {
            System.err.println("✗ Error creating table: " + e.getMessage());
            return;
        }
        
        // Insert sample products using PreparedStatement
        String insertSQL = "INSERT INTO products (name, description, price, stock_quantity, category) " +
                          "VALUES (?, ?, ?, ?, ?)";
        
        String[][] sampleProducts = {
            {"Laptop Computer", "High-performance laptop with 16GB RAM, 512GB SSD, and dedicated graphics", "999.99", "25", "Electronics"},
            {"Wireless Mouse", "Ergonomic wireless mouse with precision tracking and long battery life", "29.99", "150", "Accessories"},
            {"Mechanical Keyboard", "RGB backlit mechanical keyboard with blue switches", "89.99", "75", "Accessories"},
            {"USB-C Hub", "7-in-1 USB-C hub with HDMI, USB 3.0, and SD card reader", "49.99", "100", "Accessories"},
            {"External Hard Drive", "2TB portable external hard drive with USB 3.0", "79.99", "60", "Storage"},
            {"Laptop Bag", "Protective laptop bag with multiple compartments", "39.99", "200", "Accessories"},
            {"Webcam HD", "1080p HD webcam with built-in microphone", "59.99", "80", "Electronics"},
            {"Monitor 27-inch", "27-inch 4K monitor with HDR support", "399.99", "30", "Electronics"},
            {"Gaming Headset", "Surround sound gaming headset with noise cancellation", "79.99", "120", "Audio"},
            {"Portable SSD", "1TB portable SSD with fast read/write speeds", "129.99", "45", "Storage"}
        };
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(insertSQL)) {
            
            for (String[] product : sampleProducts) {
                pstmt.setString(1, product[0]);  // name
                pstmt.setString(2, product[1]);  // description
                pstmt.setDouble(3, Double.parseDouble(product[2]));  // price
                pstmt.setInt(4, Integer.parseInt(product[3]));  // stock
                pstmt.setString(5, product[4]);  // category
                pstmt.executeUpdate();
            }
            
            System.out.println("✓ Sample products inserted");
            System.out.println("\n✓ Setup complete!");
            System.out.println("\nTry searching for:");
            System.out.println("  - laptop");
            System.out.println("  - USB");
            System.out.println("  - gaming");
            System.out.println("  - storage");
            
        } catch (SQLException e) {
            System.err.println("✗ Error inserting products: " + e.getMessage());
        }
    }
    
    /**
     * Demonstrates SQL injection prevention.
     */
    public static void demonstrateSQLInjection() {
        System.out.println("\n=== SQL Injection Prevention Demo ===\n");
        
        String maliciousInput = "laptop' OR '1'='1";
        
        System.out.println("Malicious input: " + maliciousInput);
        System.out.println("\nVULNERABLE approach (DON'T DO THIS):");
        System.out.println("String sql = \"SELECT * FROM products WHERE name LIKE '%\" + keyword + \"%'\";");
        System.out.println("\nResulting query:");
        System.out.println("SELECT * FROM products WHERE name LIKE '%laptop' OR '1'='1%'");
        System.out.println("\nThe OR '1'='1 would return ALL products!");
        
        System.out.println("\n" + "=".repeat(70));
        System.out.println("\nSECURE approach (OUR CODE):");
        System.out.println("String sql = \"SELECT * FROM products WHERE name LIKE ?\";");
        System.out.println("pstmt.setString(1, \"%\" + keyword + \"%\");");
        System.out.println("\nHow it's handled:");
        System.out.println("Parameter value: %laptop' OR '1'='1%");
        System.out.println("\nThe entire string is treated as the search pattern!");
        System.out.println("It searches for products literally named \"laptop' OR '1'='1\"");
        System.out.println("Result: No matches found (SQL injection prevented!)");
    }
}

/*
SECURITY FEATURES:

✓ PreparedStatement with parameterized queries
✓ Input validation (length checks)
✓ Result limits (prevents memory exhaustion)
✓ Suspicious pattern detection (for logging)
✓ No string concatenation in SQL
✓ Safe LIKE pattern construction
✓ Proper error handling
✓ SQL injection prevention

HOW PARAMETERIZED QUERIES PREVENT SQL INJECTION:

VULNERABLE CODE (DON'T DO THIS):
```java
String sql = "SELECT * FROM products WHERE name LIKE '%" + keyword + "%'";
Statement stmt = conn.createStatement();
ResultSet rs = stmt.executeQuery(sql);  // DANGEROUS!
```

Attack input: laptop' OR '1'='1' --
Resulting query: SELECT * FROM products WHERE name LIKE '%laptop' OR '1'='1' --%'
Result: Returns ALL products (authentication bypass!)

SECURE CODE (OUR APPROACH):
```java
String sql = "SELECT * FROM products WHERE name LIKE ?";
PreparedStatement pstmt = conn.prepareStatement(sql);
pstmt.setString(1, "%" + keyword + "%");
ResultSet rs = pstmt.executeQuery();
```

Attack input: laptop' OR '1'='1' --
How PreparedStatement handles it:
The entire string "%laptop' OR '1'='1' --%" becomes the parameter value.
It's treated as a literal search string, not SQL code!
Result: Searches for that exact string, finds nothing (safe!)

SQL INJECTION ATTACKS PREVENTED:

1. Union-based injection:
   Input: laptop' UNION SELECT credit_card FROM users --
   Result: Treated as literal search string

2. Boolean-based injection:
   Input: laptop' AND 1=1 --
   Result: Treated as literal search string

3. Time-based injection:
   Input: laptop' AND SLEEP(5) --
   Result: Treated as literal search string

4. Stacked queries:
   Input: laptop'; DROP TABLE products; --
   Result: Treated as literal search string, table not dropped!

VALIDATION LAYERS:

1. Length Validation:
   - Minimum 1 character
   - Maximum 100 characters
   - Prevents buffer overflow attempts

2. Suspicious Pattern Detection:
   - Logs common SQL keywords
   - Doesn't block (PreparedStatement already safe)
   - Helps with security monitoring

3. Result Limiting:
   - Maximum 100 results
   - Prevents DoS via large result sets

SETUP INSTRUCTIONS:

1. Install MySQL JDBC Driver:
   Maven:
   <dependency>
       <groupId>mysql</groupId>
       <artifactId>mysql-connector-java</artifactId>
       <version>8.0.33</version>
   </dependency>

2. Create MySQL database:
   CREATE DATABASE shop_db;

3. Update connection parameters:
   - DB_URL: your database URL
   - DB_USER: your MySQL username
   - DB_PASSWORD: your MySQL password

4. Setup database (run once):
   Uncomment: setupDatabase();

5. Compile and run:
   javac SecureProductSearch.java
   java -cp .:mysql-connector-java.jar SecureProductSearch

USAGE EXAMPLES:

Search: laptop
Result: Finds "Laptop Computer", "Laptop Bag"

Search: USB
Result: Finds "USB-C Hub", "Wireless Mouse" (if USB in description)

Search: gaming
Result: Finds "Gaming Headset"

Malicious input: laptop' OR '1'='1' --
Result: No matches (SQL injection prevented!)

OUTPUT FORMAT:

ID     Name                           Price          Stock      Value
---------------------------------------------------------------------------
1      Laptop Computer                $999.99        25         $24999.75
       High-performance laptop with 16GB RAM, 512GB SSD, and dedica...

6      Laptop Bag                     $39.99         200        $7998.00
       Protective laptop bag with multiple compartments

---------------------------------------------------------------------------
Total results: 2
Total inventory value: $32997.75

PRODUCTION ENHANCEMENTS:

1. Connection pooling for better performance
2. Prepared statement caching
3. Pagination for large result sets
4. Advanced filtering (price range, category)
5. Full-text search indexes
6. Audit logging of searches
7. Rate limiting to prevent abuse
8. Cache frequently searched terms

The key principle: Always use PreparedStatement with parameterized
queries for ANY database operation involving user input!
*/